### PR TITLE
update to new observers

### PR DESCRIPTION
**Pull request by stevebachmeier**
### Description

- *Category*:  refactor
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5056
- *Research reference*:  na

### Changes and notes
 
This PR implements the new results processing system's Observers and
Observations. NOTE: it does _not_ totally remove the post-processing 
(i.e. `make_results`) step since this is a particularly unique model and
the post-processing is quite complicated.

### Verification and Testing

I did compare before/after raw (not-post-processed) results. Some differences:

- various minor dtype differences (int64 vs 32, datetime64[ms] vs [ns]
- census reproducibility: guardian_1, guardian_1_address_id
- social security: event_type dtype categorical (old) vs object (new)
- taxes 1040: alive dtype categorical (old) vs object (new)
- taxes dependents
    - reproducibility: guardian_id
    - differences old vs new: copy_age, copy_date_of_birth, copy_ssn
- wic reproducibility: simulant_id (!), guardian_id

I'm reasonably convinced that the differences are due to reproducibility issues which is out of scope for this ticket. Notebook I'm using for these checks can be found at /mnt/homes/sbachmei/notebooks/simscience/results-processing/vv-census-raw-results.ipynb (cannot attach .ipynb)

### CODE REVIEW

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 1]
```diff
@@ -233,31 +164,41 @@ def setup(self, builder: Builder):
...
-    ########################
+    def get_configuration_name(self) -> str:
+        return "_".join(self.name.split("_observer_"))
+
+    def register_observations(self, builder: Builder) -> None:
+        # While this is a concatenating observation, it is not registered as such
+        # because we need to modify the per-time-step results_gatherer method
+        # FIXME: Is tracked==True correct? We probably want to exclude households
+        # that are entirely untracked, but should we allow for simulants
+        # to, say, copy from household members that are now untracked?
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/components/observers.py:175]: This is a good question. Im pretty sure the copy from household member looks at the household to see if there are members to copy so I would guess you can't copy an untracked simulant but testing that would be bettert han taking my word for it.

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 2]
```diff
@@ -338,20 +286,29 @@ def setup(self, builder: Builder):
...
-            event.index,
-            query="alive == 'alive'",  # WIC should include only living simulants
-        )
+    def get_observation(self, pop: pd.DataFrame) -> pd.DataFrame:
         pop["income"] = self.pipelines["income"](pop.index)
 
         # add columns for output
-        pop["year"] = event.time.year
+        pop["simulant_id"] = pop.index
+        pop["year"] = pop["event_time"].dt.year
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/components/observers.py:311]: There is some sort of helper function you could put here and all the other places you need to get random pop columns to reduce some of the duplicated code. Maybe its more trouble than its worth but there is somethign there.

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 3]
```diff
@@ -411,28 +368,27 @@ def get_observation(self, event: Event) -> pd.DataFrame:
...
-                        race_eth_rows
-                    ],  # keep pr of 1.0 for the k infants with mother on WIC
-                    (pr_covered[race_ethnicity] * N - k)
-                    / (N - k)  # rescale probability for the remaining individuals
-                    # so that expected number of infants on WIC matches target
+                    child_covered_pr[race_eth_rows],
+                    (pr_covered[race_ethnicity] * num_infants_total - num_infants_included)
+                    / (num_infants_total - num_infants_included),
                 )
         pop_included[0] = self.randomness.filter_for_probability(pop_u1, child_covered_pr)
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/components/observers.py:393]: pop_included is a dict and should always be the same order but will the groupby object also always be ordered the same?

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 4]
```diff
@@ -908,27 +913,41 @@ def setup(self, builder: Builder):
         self.clock = builder.time.clock()
         self.randomness = builder.randomness.get_stream(self.name)
 
+    def register_observations(self, builder: Builder) -> None:
+        # While this is a concatenating observation, it is not registered as such
+        # because we need to modify the per-time-step results_gatherer method
+        builder.results.register_unstratified_observation(
+            name=self.output_name,
+            pop_filter="",  # want to include untracked simulants
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/components/observers.py:921]: I feel like this should be the default for this arg in vivarium.

**Review by albrja — APPROVED** [review]

File: src/vivarium_census_prl_synth_pop/components/observers.py (hunk 4 above)
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/components/observers.py:921]: We need to have a larger discussion about changing fundamental defaults like this. I don't think any of us like the whole tracked thing anyway...

File: src/vivarium_census_prl_synth_pop/components/observers.py (hunk 3 above)
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/components/observers.py:393]: I'm not sure I understand the question here. pop_included is a dict where keys are the ages. This line is just adding the values for age 0.

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 5]
```diff
@@ -479,16 +436,34 @@ def setup(self, builder: Builder):
...
         self.end_time = dt.datetime(**builder.configuration.time.end)
 
+    def register_observations(self, builder: Builder) -> None:
+        # builder.results.register_concatenating_observation(
+        #     name=self.output_name,
+        #     pop_filter="has_ssn == True",
+        #     requires_columns=self.columns_required,
+        #     results_formatter=self.format,
+        #     to_observe=self.to_observe,
+        # )
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/observers.py:446]: Remove this?

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 6]
```diff
@@ -479,16 +436,34 @@ def setup(self, builder: Builder):
...
         return self.clock() < self.end_time <= event.time
 
-    def get_observation(self, event: Event) -> pd.DataFrame:
-        pop = self.population_view.get(
-            event.index,
-            query="has_ssn == True",  # only include simulants with a SSN
-        )
+    # def format(self, measure: str, results: pd.DataFrame) -> pd.DataFrame:
+    def get_observation(self, pop: pd.DataFrame) -> pd.DataFrame:
+        # pop = results.copy()
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/observers.py:464]: Remove these commented lines?

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 7]
```diff
@@ -479,16 +436,34 @@ def setup(self, builder: Builder):
         self.start_time = dt.datetime(**builder.configuration.time.start)
         self.end_time = dt.datetime(**builder.configuration.time.end)
 
+    def register_observations(self, builder: Builder) -> None:
+        # builder.results.register_concatenating_observation(
```
Comment (patricktnast) [inline src/vivarium_census_prl_synth_pop/components/observers.py:440]: intend to keep as commented out?

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 8]
```diff
@@ -479,16 +436,34 @@ def setup(self, builder: Builder):
...
     def to_observe(self, event: Event) -> bool:
         """Only observe if this is the final time step of the sim"""
         return self.clock() < self.end_time <= event.time
 
-    def get_observation(self, event: Event) -> pd.DataFrame:
-        pop = self.population_view.get(
-            event.index,
-            query="has_ssn == True",  # only include simulants with a SSN
-        )
+    # def format(self, measure: str, results: pd.DataFrame) -> pd.DataFrame:
```
Comment (patricktnast) [inline src/vivarium_census_prl_synth_pop/components/observers.py:462]: intend to keep as commented out?

File: src/vivarium_census_prl_synth_pop/components/observers.py [hunk 9]
```diff
@@ -908,27 +913,41 @@ def setup(self, builder: Builder):
...
+        )
+
     def to_observe(self, event: Event) -> bool:
         """Observe if April 15 falls during this time step"""
         tax_date = dt.datetime(event.time.year, 4, 15)
         return self.clock() < tax_date <= event.time
 
-    def get_observation(self, event: Event) -> pd.DataFrame:
-        pop = self.population_view.get(event.index)
+    # def get_observation(self, event: Event) -> pd.DataFrame:
```
Comment (patricktnast) [inline src/vivarium_census_prl_synth_pop/components/observers.py:933]: intend to keep as commented out?

File: src/vivarium_census_prl_synth_pop/components/observers.py
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/observers.py:1]: You're not consistent with regard to copying the `pop` dataframe. Is this correct?

**Review by patricktnast — APPROVED** [review]
I'm not super familiar with some of the details, but the general observers implementation LGTM. There is some commented out code scattered throughout that I don't know if you meant to keep in.

**Review by rmudambi — APPROVED** [review]

File: pyproject.toml [hunk 10]
```diff
@@ -9,3 +9,5 @@ show_missing = true
 
 [tool.isort]
 profile = "black"
+skip = ["vivarium", "vivarium_public_health"]
+known_third_party = ["vivarium", "vivarium_public_health"]
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply pyproject.toml:13]: Not really, other than to say that we really don't want to run isort on external packages. I'm unsure why it's failing to begin, though. We do do this elsewhere as needed as well

File: tests/components/observers/test_household_survey_observers.py [hunk 11]
```diff
@@ -112,11 +97,11 @@ def test_multiple_observation(
...
     """Are responses recorded correctly (including concatenation after time steps)?"""
-    sim_start_date = "2021-01-01"
+    sim_start_date = str(mocked_pop_view.event_time.iat[0].date())
     # Simulate first time step
-    event = mocker.MagicMock()
-    event.time = pd.to_datetime(sim_start_date)
-    event.index = pd.Index([0, 1])
+    # event = mocker.MagicMock()
+    # event.time = pd.to_datetime(sim_start_date)
+    # event.index = pd.Index([0, 1])
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply tests/components/observers/test_household_survey_observers.py:104]: Oops, just forgot to delete (we don't use event anymore here)

File: src/vivarium_census_prl_synth_pop/utilities.py [hunk 12]
```diff
@@ -490,10 +477,11 @@ def copy_from_household_member(
 ) -> pd.DataFrame:
     # Creates copy_age, copy_date_of_birth, and copy_ssn from household members
     # Note: copy value can be original value but copies from another household member
-
+    pop = pop.copy()
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/utilities.py:480]: I don't recall why I felt the need to copy, but I agree we don't

**Review by albrja — APPROVED** [review]

File: src/vivarium_census_prl_synth_pop/utilities.py [hunk 13]
```diff
@@ -472,12 +471,19 @@ def write_to_disk(data: pd.DataFrame, path: Path):
         )
 
 
+def convert_objects_to_categories(data: pd.DataFrame) -> pd.DataFrame:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/utilities.py:474]: I separated out this portion b/c vivarium is now writing out results inherently and so I use this function as the observation.results_formatter method.

**Review by patricktnast — APPROVED** [review]

**Pull request merged**